Mathlib/Algebra/Group/Action/Equidecomp.lean

```diff
@@ -0,0 +1,240 @@
+-/
+structure _root_.Equidecomp extends PartialEquiv X X where
+  decomp' : ∃ S : Finset G, DecompOn toFun source S
+
+variable {X G}
+
+/-- Note that `Equidecomp X G` is not `FunLike`. -/
+instance : CoeFun (Equidecomp X G) fun _ => X → X := ⟨fun f => f.toFun⟩
+
+/-- A finite set of group elements witnessing that `f` is an equidecomposition. -/
+noncomputable
+def elements (f : Equidecomp X G) := f.decomp'.choose
+
+theorem decomp (f : Equidecomp X G) : DecompOn f f.source f.elements := f.decomp'.choose_spec
+
+@[simp]
+theorem map_source {f : Equidecomp X G} {x : X} (h : x ∈ f.source) :
+    f x ∈ f.target := f.toPartialEquiv.map_source h
+
+private theorem ext' {f f' : Equidecomp X G} (h_equiv : f.toPartialEquiv = f'.toPartialEquiv) :
+    f = f' := by
+  let ⟨_, _, _⟩ := f
+  congr
+
+theorem DecompOn.mono {f f' : X → X} {A A' : Set X} {S : Finset G} (h : DecompOn f A S)
+    (hA' : A' ⊆ A) (hf' : EqOn f f' A') : DecompOn f' A' S := by
+  intro a ha
+  rcases h _ (hA' ha) with ⟨γ, γ_mem, hγ⟩
+  use γ, γ_mem
+  rwa [← hf' ha]
+
+/-- The restriction of an equidecomposition is an equidecomposition. -/
+def restr (f : Equidecomp X G) (A : Set X) : Equidecomp X G where
+  toPartialEquiv := f.toPartialEquiv.restr A
+  decomp' := ⟨f.elements,
+    f.decomp.mono (restr_source_subset_source _ _) (by exact fun ⦃x⦄ ↦ congrFun rfl)⟩
+
+@[simp]
+theorem restr_toPartialEquiv (f : Equidecomp X G) (A : Set X) :
+    (f.restr A).toPartialEquiv = f.toPartialEquiv.restr A := rfl
+
+theorem restr_source (f : Equidecomp X G) (A : Set X) : (f.restr A).source = f.source ∩ A := rfl
+
+theorem restr_source_of_subset (f : Equidecomp X G) {A : Set X} (hA : A ⊆ f.source) :
+    (f.restr A).source = A := by rw [restr_source, inter_eq_self_of_subset_right hA]
+
+theorem restr_eq_of_source_subset {f : Equidecomp X G} {A : Set X} (hA : f.source ⊆ A) :
+    f.restr A = f :=
+  Equidecomp.ext' (by rw [restr_toPartialEquiv, PartialEquiv.restr_eq_of_source_subset hA])
+
+@[simp]
+theorem restr_univ (f : Equidecomp X G) : f.restr univ = f :=
+  restr_eq_of_source_subset <| subset_univ _
+
+end SMul
+
+open scoped Classical
+
+section Monoid
+
+variable [Monoid G] [MulAction G X]
+
+variable (X G)
+
+/-- The identity function is an equidecomposition of the space with itself. -/
+def refl : Equidecomp X G where
+  toPartialEquiv := PartialEquiv.refl _
+  decomp' := ⟨{1}, by simp [DecompOn]⟩
+
+@[simp]
+theorem refl_toPartialEquiv : (Equidecomp.refl X G).toPartialEquiv = PartialEquiv.refl X := rfl
+
+variable {X}
+
+/-- The identity function is an equidecomposition of any set with itself. -/
+def reflSet (A : Set X) : Equidecomp X G := (Equidecomp.refl X G).restr A
+
+theorem reflSet_source (A : Set X) : (reflSet G A).source = A :=
+  restr_source_of_subset _ <| subset_univ _
+
+@[simp]
+theorem reflSet_toPartialEquiv (A : Set X) :
+    (reflSet G A).toPartialEquiv = PartialEquiv.ofSet A := by
+  ext; rfl; rfl; rw [reflSet_source, ofSet_source]
```

Comment:
linter complains about using rfl without being explicit about the goal: try rfl on goals and rewrite on the last one.

```suggestion
  ext <;> try rfl
  rw [reflSet_source, ofSet_source]
```

Reply:
Note that this is the error that breaks CI.